File: 53-Maximum-Subarray-MEDIUM/kadane.scala
object Solution {
    def maxSubArray(nums: Array[Int]): Int = {
        nums.foldLeft((0,Int.MinValue)) {
            case ((currSum, maxSum), elem) =>
                val newCurrSum = Math.max(elem, currSum + elem)
                val newMaxSum = Math.max(maxSum, newCurrSum)
                (newCurrSum, newMaxSum)
        }._2
    }
}

File: 53-Maximum-Subarray-MEDIUM/kadane.go
func max(a, b int) int {
	if a < b {
		return b
	} else {
		return a
	}
}

func maxSubArray(nums []int) int {
	currentSum := nums[0]
	maxSum := nums[0]

	for _, v := range nums[1:] {
		currentSum = max(currentSum+v, v)
		maxSum = max(maxSum, currentSum)
	}

	return maxSum
}
